File: src/services/support.service.ts
/**
 * Support Service
 * Business logic for support ticket system
 */

import { AppDataSource } from '../database';
import { SupportTicket, SupportMessage } from '../database/entities';
import { MoreThan } from 'typeorm';
import type { SupportCategory, SupportStatus, SupportSender } from '../database/entities/SupportTicket.entity';
import { AdminSession } from '../database/entities/AdminSession.entity';

export interface CreateTicketData {
  userId: number;
  category: SupportCategory;
  initialMessage?: string;
  attachments?: Array<{ type: string; file_id: string; caption?: string }>;
}

export interface AddMessageData {
  ticketId: number;
  sender: SupportSender;
  adminId?: number;
  text?: string;
  attachments?: Array<{ type: string; file_id: string; caption?: string }>;
}

export class SupportService {
  private ticketRepo = AppDataSource.getRepository(SupportTicket);
  private messageRepo = AppDataSource.getRepository(SupportMessage);
  private adminSessionRepo = AppDataSource.getRepository(AdminSession);

  /**
   * Create a new support ticket
   */
  async createTicket(data: CreateTicketData): Promise<SupportTicket> {
    // Check if user already has an open ticket
    const existingOpen = await this.ticketRepo.findOne({
      where: {
        user_id: data.userId,
        status: 'open' as SupportStatus,
      },
    });

    if (existingOpen) {
      throw new Error('У вас уже есть открытое обращение. Пожалуйста, дождитесь ответа администратора.');
    }

    // Check in_progress and answered as well (partial UNIQUE index covers these)
    const existingActive = await this.ticketRepo
      .createQueryBuilder('ticket')
      .where('ticket.user_id = :userId', { userId: data.userId })
      .andWhere('ticket.status IN (:...statuses)', { statuses: ['open', 'in_progress', 'answered'] })
      .getOne();

    if (existingActive) {
      throw new Error('У вас уже есть активное обращение. Пожалуйста, дождитесь его закрытия.');
    }

    // Create ticket
    const ticket = this.ticketRepo.create({
      user_id: data.userId,
      category: data.category,
      status: 'open',
      last_user_message_at: new Date(),
    });

    await this.ticketRepo.save(ticket);

    // Add initial message if provided
    if (data.initialMessage || data.attachments) {
      await this.addUserMessage({
        ticketId: ticket.id,
        sender: 'user',
        text: data.initialMessage,
        attachments: data.attachments,
      });
    }

    return ticket;
  }

  /**
   * Add user message to ticket
   */
  async addUserMessage(data: AddMessageData): Promise<SupportMessage> {
    const message = this.messageRepo.create({
      ticket_id: data.ticketId,
      sender: 'user',
      text: data.text,
      attachments: data.attachments,
    });

    await this.messageRepo.save(message);

    // Update ticket last_user_message_at
    await this.ticketRepo.update(data.ticketId, {
      last_user_message_at: new Date(),
      status: 'open', // Reset to open when user sends new message
    });

    return message;
  }

  /**
   * Add admin message to ticket
   */
  async addAdminMessage(data: AddMessageData): Promise<SupportMessage> {
    if (!data.adminId) {
      throw new Error('Admin ID required for admin messages');
    }

    const message = this.messageRepo.create({
      ticket_id: data.ticketId,
      sender: 'admin',
      admin_id: data.adminId,
      text: data.text,
      attachments: data.attachments,
    });

    await this.messageRepo.save(message);

    // Update ticket last_admin_message_at and status
    await this.ticketRepo.update(data.ticketId, {
      last_admin_message_at: new Date(),
      status: 'answered', // Mark as answered when admin replies
    });

    return message;
  }

  /**
   * Add system message to ticket
   */
  async addSystemMessage(ticketId: number, text: string): Promise<SupportMessage> {
    const message = this.messageRepo.create({
      ticket_id: ticketId,
      sender: 'system',
      text,
    });

    await this.messageRepo.save(message);
    return message;
  }

  /**
   * Assign ticket to admin
   */
  async assignToSelf(ticketId: number, adminId: number): Promise<void> {
    await this.ticketRepo.update(ticketId, {
      assigned_admin_id: adminId,
      status: 'in_progress',
    });
  }

  /**
   * Close ticket
   */
  async close(ticketId: number): Promise<void> {
    await this.ticketRepo.update(ticketId, {
      status: 'closed',
    });
  }

  /**
   * Reopen closed ticket
   */
  async reopen(ticketId: number): Promise<void> {
    await this.ticketRepo.update(ticketId, {
      status: 'open',
    });
  }

  /**
   * List open tickets for admin
   */
  async listOpen(): Promise<SupportTicket[]> {
    return this.ticketRepo.find({
      where: [
        { status: 'open' },
        { status: 'in_progress' },
        { status: 'answered' },
      ],
      relations: ['user', 'assigned_admin'],
      order: {
        created_at: 'DESC',
      },
    });
  }

  /**
   * Get ticket by ID with messages
   */
  async get(ticketId: number): Promise<SupportTicket | null> {
    return this.ticketRepo.findOne({
      where: { id: ticketId },
      relations: ['user', 'assigned_admin', 'messages'],
      order: {
        messages: {
          created_at: 'ASC',
        },
      },
    });
  }

  /**
   * Get user's active ticket (open, in_progress, or answered)
   */
  async getUserActiveTicket(userId: number): Promise<SupportTicket | null> {
    return this.ticketRepo
      .createQueryBuilder('ticket')
      .where('ticket.user_id = :userId', { userId })
      .andWhere('ticket.status IN (:...statuses)', { statuses: ['open', 'in_progress', 'answered'] })
      .leftJoinAndSelect('ticket.messages', 'messages')
      .orderBy('messages.created_at', 'ASC')
      .getOne();
  }

  /**
   * Find on-duty admin (admin with active session)
   */
  async findOnDutyAdmin(): Promise<number | null> {
    const now = new Date();

    // Find active admin session
    const activeSession = await this.adminSessionRepo.findOne({
      where: {
        is_active: true,
        expires_at: MoreThan(now),
      },
      order: {
        last_activity: 'DESC',
      },
    });

    return activeSession?.admin_id || null;
  }
}

export const supportService = new SupportService();

File: app/services/support_service.py
"""
Support service.

Manages support ticket system for users.
"""

from datetime import UTC, datetime
from typing import Any

from loguru import logger
from sqlalchemy import or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.enums import SupportSender, SupportStatus
from app.models.support_message import SupportMessage
from app.models.support_ticket import SupportTicket
from app.repositories.support_message_repository import (
    SupportMessageRepository,
)
from app.repositories.support_ticket_repository import (
    SupportTicketRepository,
)


class SupportService:
    """Support service for ticket management."""

    def __init__(self, session: AsyncSession) -> None:
        """Initialize support service."""
        self.session = session
        self.ticket_repo = SupportTicketRepository(session)
        self.message_repo = SupportMessageRepository(session)

    async def create_ticket(
        self,
        user_id: int | None,
        category: str,
        initial_message: str | None = None,
        attachments: list[dict[str, Any]] | None = None,
        telegram_id: int | None = None,
    ) -> tuple[SupportTicket | None, str | None]:
        """
        Create new support ticket.

        Args:
            user_id: User ID (None for guest tickets)
            category: Ticket category
            initial_message: Initial message text
            attachments: List of attachment dicts
            telegram_id: Telegram ID (required for guest tickets when user_id is None)

        Returns:
            Tuple of (ticket, error_message)
        """
        # Validate: if user_id is None, telegram_id must be provided
        if user_id is None and telegram_id is None:
            return None, "Для гостевых тикетов требуется telegram_id"

        # Check total open tickets limit (removed single active ticket check)
        from app.config.constants import MAX_OPEN_TICKETS_PER_USER
        from app.models.enums import SupportTicketStatus

        if user_id is not None:
            # Count open tickets for registered user
            open_tickets = await self.ticket_repo.get_by_user(
                user_id, status=SupportTicketStatus.OPEN.value
            )
        else:
            # Count open tickets for guest
            all_guest_tickets = await self.ticket_repo.get_by_telegram_id(
                telegram_id
            )
            open_tickets = [
                t
                for t in all_guest_tickets
                if t.status == SupportTicketStatus.OPEN.value
            ]

        if len(open_tickets) >= MAX_OPEN_TICKETS_PER_USER:
            return None, (
                f"Превышен лимит открытых обращений "
                f"({MAX_OPEN_TICKETS_PER_USER}). "
                "Пожалуйста, дождитесь закрытия существующих обращений."
            )

        # Create ticket
        ticket = await self.ticket_repo.create(
            user_id=user_id,
            telegram_id=telegram_id,
            category=category,
            status=SupportStatus.OPEN.value,
            last_user_message_at=datetime.now(UTC),
        )

        # Add initial message if provided
        if initial_message or attachments:
            await self.add_user_message(
                ticket_id=ticket.id,
                text=initial_message,
                attachments=attachments,
            )

        await self.session.commit()

        logger.info(
            "Support ticket created",
            extra={
                "ticket_id": ticket.id,
                "user_id": user_id,
                "telegram_id": telegram_id,
                "category": category,
                "is_guest": user_id is None,
            },
        )

        return ticket, None

    async def add_user_message(
        self,
        ticket_id: int,
        text: str | None = None,
        attachments: list[dict[str, Any]] | None = None,
    ) -> SupportMessage:
        """
        Add user message to ticket.

        Args:
            ticket_id: Ticket ID
            text: Message text
            attachments: List of attachment dicts

        Returns:
            SupportMessage record
        """
        # Create message
        message = await self.message_repo.create(
            ticket_id=ticket_id,
            sender=SupportSender.USER.value,
            text=text,
            attachments=attachments,
        )

        # Update ticket timestamp and reset to open
        await self.ticket_repo.update(
            ticket_id,
            last_user_message_at=datetime.now(UTC),
            status=SupportStatus.OPEN.value,
        )

        await self.session.flush()

        return message

    async def add_admin_message(
        self,
        ticket_id: int,
        admin_id: int,
        text: str | None = None,
        attachments: list[dict[str, Any]] | None = None,
    ) -> SupportMessage:
        """
        Add admin message to ticket.

        Args:
            ticket_id: Ticket ID
            admin_id: Admin ID
            text: Message text
            attachments: List of attachment dicts

        Returns:
            SupportMessage record
        """
        # Create message
        message = await self.message_repo.create(
            ticket_id=ticket_id,
            sender=SupportSender.ADMIN.value,
            admin_id=admin_id,
            text=text,
            attachments=attachments,
        )

        # Update ticket timestamp and mark as answered
        await self.ticket_repo.update(
            ticket_id,
            last_admin_message_at=datetime.now(UTC),
            status=SupportStatus.ANSWERED.value,
        )

        await self.session.flush()

        return message

    async def add_system_message(
        self, ticket_id: int, text: str
    ) -> SupportMessage:
        """
        Add system message to ticket.

        Args:
            ticket_id: Ticket ID
            text: Message text

        Returns:
            SupportMessage record
        """
        message = await self.message_repo.create(
            ticket_id=ticket_id,
            sender=SupportSender.SYSTEM.value,
            text=text,
        )

        await self.session.flush()

        return message

    async def assign_to_admin(
        self, ticket_id: int, admin_id: int
    ) -> None:
        """
        Assign ticket to admin.

        Args:
            ticket_id: Ticket ID
            admin_id: Admin ID
        """
        await self.ticket_repo.update(
            ticket_id,
            assigned_admin_id=admin_id,
            status=SupportStatus.IN_PROGRESS.value,
        )

        await self.session.flush()

        logger.info(
            "Ticket assigned to admin",
            extra={"ticket_id": ticket_id, "admin_id": admin_id},
        )

    async def close_ticket(self, ticket_id: int) -> None:
        """
        Close ticket.

        Args:
            ticket_id: Ticket ID
        """
        await self.ticket_repo.update(
            ticket_id, status=SupportStatus.CLOSED.value
        )

        await self.session.flush()

        logger.info("Ticket closed", extra={"ticket_id": ticket_id})

    async def reopen_ticket(self, ticket_id: int) -> None:
        """
        Reopen closed ticket.

        Args:
            ticket_id: Ticket ID
        """
        await self.ticket_repo.update(
            ticket_id, status=SupportStatus.OPEN.value
        )

        await self.session.flush()

        logger.info("Ticket reopened", extra={"ticket_id": ticket_id})

    async def list_open_tickets(self) -> list[SupportTicket]:
        """
        List all open tickets (for admin).

        Returns:
            List of open/in_progress/answered tickets
        """
        stmt = (
            select(SupportTicket)
            .where(
                or_(
                    SupportTicket.status == SupportStatus.OPEN.value,
                    SupportTicket.status == SupportStatus.IN_PROGRESS.value,
                    SupportTicket.status == SupportStatus.ANSWERED.value,
                )
            )
            .order_by(SupportTicket.created_at.desc())
        )

        result = await self.session.execute(stmt)
        return list(result.scalars().all())

    async def get_ticket_by_id(
        self, ticket_id: int
    ) -> SupportTicket | None:
        """
        Get ticket by ID with messages.

        Args:
            ticket_id: Ticket ID

        Returns:
            SupportTicket or None
        """
        ticket = await self.ticket_repo.get_by_id(ticket_id)

        if ticket:
            # Load messages separately and sort
            messages = await self.message_repo.find_by(
                ticket_id=ticket_id
            )
            messages.sort(key=lambda m: m.created_at)
            ticket.messages = messages  # type: ignore

        return ticket

    async def get_user_tickets(
        self, user_id: int
    ) -> list[SupportTicket]:
        """
        Get all tickets for user.

        Args:
            user_id: User ID

        Returns:
            List of user's tickets
        """
        return await self.ticket_repo.get_by_user(user_id)

    async def get_user_active_ticket(
        self, user_id: int
    ) -> SupportTicket | None:
        """
        Get user's active ticket if exists.

        Args:
            user_id: User ID

        Returns:
            Active ticket or None
        """
        return await self.ticket_repo.get_active_by_user(user_id)

    async def get_guest_tickets(
        self, telegram_id: int
    ) -> list[SupportTicket]:
        """
        Get all tickets for guest by telegram_id.

        Args:
            telegram_id: Telegram ID

        Returns:
            List of guest tickets (all statuses)
        """
        return await self.ticket_repo.get_by_telegram_id(telegram_id)

    async def get_tickets_by_admin(self, admin_id: int) -> list[SupportTicket]:
        """
        Get tickets assigned to admin.

        Args:
            admin_id: Admin ID

        Returns:
            List of tickets assigned to admin
        """
        stmt = (
            select(SupportTicket)
            .where(
                SupportTicket.assigned_admin_id == admin_id,
                SupportTicket.status != SupportStatus.CLOSED.value,
            )
            .order_by(SupportTicket.updated_at.desc())
        )
        result = await self.session.execute(stmt)
        return list(result.scalars().all())

    async def get_support_stats(self) -> dict[str, int]:
        """
        Get support system statistics.

        Returns:
            Dict with statistics (total, open, closed, etc.)
        """
        from sqlalchemy import func

        stats = {
            "total": 0,
            "open": 0,
            "in_progress": 0,
            "closed": 0,
            "waiting_user": 0,
        }

        # Count total
        total_stmt = select(func.count(SupportTicket.id))
        stats["total"] = (await self.session.execute(total_stmt)).scalar() or 0

        # Count by status
        status_stmt = select(
            SupportTicket.status, func.count(SupportTicket.id)
        ).group_by(SupportTicket.status)
        
        status_results = (await self.session.execute(status_stmt)).all()
        
        for status, count in status_results:
            if status == SupportStatus.OPEN.value:
                stats["open"] = count
            elif status == SupportStatus.IN_PROGRESS.value:
                stats["in_progress"] = count
            elif status == SupportStatus.CLOSED.value:
                stats["closed"] = count
            elif status == SupportStatus.WAITING_USER.value:
                stats["waiting_user"] = count
                
        return stats